File: app/static/js/fp_triage.js
// WAF false-positive explainer — header Tools menu:  FWFpTriage.open()
// Paste an attack-log line (syslog key=value, JSON, or a raw HTTP request) and
// get: which module blocked it, which carve-out type addresses that module, the
// fields that scope it, and the exact FortiWeb payload — proved by running the
// real assembly, not a friendlier copy of it.
//
// Light chrome ONLY (docs/safeguards.md §9m). CSP-safe: no inline handlers.
(function () {
  if (window.FWFpTriage) return;

  function esc(s) {
    return String(s === undefined || s === null ? '' : s).replace(/[&<>"']/g, c =>
      ({ '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;', "'": '&#39;' }[c]));
  }
  function $(id) { return document.getElementById(id); }

  const MODAL_ID = 'fw-fptriage';
  let built = false;

  const SAMPLE = 'date=2026-08-16 time=11:20:03 log_id="20000010" msg_id=000000123456 ' +
    'device_id=FVVM00 vd="root" policy="pol-shop-cms" main_type="Signature Detection" ' +
    'sub_type="Cross Site Scripting" action="Alert_Deny" severity_level="High" ' +
    'src=198.51.100.31 src_port=51422 dst=192.0.2.90 dst_port=443 ' +
    'http_method="post" http_url="/api/v2/tickets?draft=1" http_host="soporte.example.com" ' +
    'http_agent="Mozilla/5.0" signature_id="090200001" ' +
    'msg="Cross Site Scripting detected in ARGS:body"';

  function buildModal() {
    if (built) return;
    built = true;
    const wrap = document.createElement('div');
    wrap.className = 'modal fade';
    wrap.id = MODAL_ID;
    wrap.tabIndex = -1;
    wrap.innerHTML =
      '<div class="modal-dialog modal-xl modal-dialog-scrollable"><div class="modal-content">' +
      '<div class="modal-header">' +
      '<h5 class="modal-title"><i class="bi bi-shield-exclamation me-2"></i>False-positive explainer</h5>' +
      '<button type="button" class="btn-close" data-bs-dismiss="modal"></button></div>' +
      '<div class="modal-body">' +
      '<div class="row g-2">' +
      '<div class="col-md-9">' +
      '<label class="form-label small text-muted mb-1">Attack-log entry, or a raw HTTP request</label>' +
      '<textarea id="fw-fp-in" class="form-control font-monospace" rows="6" ' +
      'placeholder="date=… main_type=&quot;Signature Detection&quot; http_url=&quot;/x&quot; signature_id=&quot;0902…&quot;&#10;— or —&#10;{&quot;main_type&quot;: &quot;…&quot;}&#10;— or —&#10;POST /api/v2/tickets HTTP/1.1"></textarea>' +
      '</div>' +
      '<div class="col-md-3 d-flex flex-column gap-2 justify-content-end">' +
      '<button class="fw-btn fw-btn-primary" id="fw-fp-run"><i class="bi bi-search me-1"></i>Explain</button>' +
      '<button class="fw-btn" id="fw-fp-sample"><i class="bi bi-file-text me-1"></i>Load a sample</button>' +
      '<button class="fw-btn" id="fw-fp-clear"><i class="bi bi-x-lg me-1"></i>Clear</button>' +
      '</div></div>' +
      '<div id="fw-fp-err" class="text-danger small mt-2"></div>' +
      '<div id="fw-fp-out" class="mt-3"></div>' +
      '</div></div></div>';
    document.body.appendChild(wrap);
    $('fw-fp-run').addEventListener('click', run);
    $('fw-fp-sample').addEventListener('click', function () { $('fw-fp-in').value = SAMPLE; run(); });
    $('fw-fp-clear').addEventListener('click', function () {
      $('fw-fp-in').value = ''; $('fw-fp-out').innerHTML = ''; $('fw-fp-err').textContent = '';
    });
  }

  function post(url, body) {
    const opts = {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', 'Accept': 'application/json' },
      body: JSON.stringify(body)
    };
    const meta = document.querySelector('meta[name="csrf-token"]');
    if (meta && meta.content) opts.headers['X-CSRFToken'] = meta.content;
    return fetch(url, opts).then(r => r.json().then(d => ({ status: r.status, d: d })));
  }

  function run() {
    const text = $('fw-fp-in').value;
    if (!text.trim()) { $('fw-fp-err').textContent = 'Paste an entry first.'; return; }
    $('fw-fp-err').textContent = '';
    $('fw-fp-out').innerHTML = '<div class="text-muted small">' +
      '<span class="spinner-border spinner-border-sm me-2"></span>Reading the entry…</div>';
    post('/fp-triage/triage', { text: text }).then(r => {
      if (!r.d || !r.d.ok) {
        $('fw-fp-out').innerHTML = '';
        $('fw-fp-err').textContent = (r.d && r.d.error) || 'Could not read that.';
        return;
      }
      render(r.d);
    }).catch(e => { $('fw-fp-err').textContent = String(e); });
  }

  function kv(k, v) {
    return '<div class="d-flex justify-content-between border-bottom py-1 small">' +
      '<span class="text-muted">' + esc(k) + '</span>' +
      '<span class="font-monospace text-end">' + esc(v) + '</span></div>';
  }

  function render(d) {
    let h = '';

    // --- what was read, and what was NOT --------------------------------
    h += '<div class="fw-card p-3 mb-3">' +
      '<div class="d-flex justify-content-between align-items-center mb-2">' +
      '<strong>Entry read</strong><span class="fw-badge fw-badge-secondary">' +
      esc(d.fmt || 'unknown format') + '</span></div>';
    Object.keys(d.row || {}).forEach(k => { h += kv(k, d.row[k]); });
    if (d.mapped && Object.keys(d.mapped).length) {
      h += '<div class="small text-muted mt-2">Renamed to SATOM\'s field names: ' +
        esc(Object.keys(d.mapped).map(k => k + ' → ' + d.mapped[k]).join(', ')) + '</div>';
    }
    if (d.unmapped && Object.keys(d.unmapped).length) {
      // Not debris: a field SATOM could not place is evidence the
      // recommendation below never saw.
      h += '<div class="alert alert-warning py-2 small mt-2 mb-0">' +
        '<i class="bi bi-exclamation-triangle me-1"></i><strong>Not used:</strong> ' +
        esc(Object.keys(d.unmapped).join(', ')) +
        ' — these keys are not attack-log fields SATOM recognises, so nothing below rests on them.</div>';
    }
    h += '</div>';

    if (d.missing && d.missing.length) {
      h += '<div class="alert alert-info py-2 small mb-3"><strong>Absent from this entry, and what each one decides:</strong><ul class="mb-0 mt-1">';
      d.missing.forEach(m => { h += '<li><span class="font-monospace">' + esc(m[0]) + '</span> — ' + esc(m[1]) + '</li>'; });
      h += '</ul></div>';
    }

    if (d.decoded && d.decoded.length) {
      h += '<div class="fw-card p-3 mb-3"><div class="fw-semibold mb-2">Payload, decoded</div>';
      d.decoded.forEach(l => {
        h += '<div class="small py-1 border-bottom"><span class="fw-badge fw-badge-info me-2">' +
          esc(l.how) + '</span><span class="font-monospace">' + esc(l.value) + '</span></div>';
      });
      h += '</div>';
    }

    // --- the carve-outs -------------------------------------------------
    (d.types || []).forEach((t, i) => {
      const rec = t.recommended || {};
      const prev = rec.preview || {};
      h += '<div class="fw-card p-3 mb-2">' +
        '<div class="d-flex justify-content-between align-items-start mb-1">' +
        '<div><strong>' + esc(t.label) + '</strong>' +
        '<div class="small text-muted">' + esc(t.group || '') + '</div></div>' +
        '<span class="fw-badge ' + (i === 0 ? 'fw-badge-success' : 'fw-badge-secondary') + '">' +
        (i === 0 ? 'recommended' : 'alternative') + '</span></div>' +
        '<div class="small mb-2">' + esc(t.why) + '</div>';

      if (t.subject && t.subject.value) {
        h += '<div class="small mb-2"><i class="bi bi-info-circle me-1"></i>' +
          esc(t.subject.label) + ' <span class="font-monospace">' + esc(t.subject.value) +
          '</span> — ' + esc(t.subject.note) + '</div>';
      }
      if ((t.scopers || []).length) {
        h += '<div class="small fw-semibold mt-2">Scoped by</div>';
        t.scopers.forEach(s => {
          const on = (rec.picked || []).indexOf(s.row_key) >= 0;
          h += '<div class="small py-1 border-bottom">' +
            '<span class="fw-badge ' + (on ? 'fw-badge-success' : 'fw-badge-secondary') + ' me-2">' +
            (on ? 'selected' : (s.required ? 'REQUIRED, absent' : 'available')) + '</span>' +
            esc(s.label) + (s.value ? ' = <span class="font-monospace">' + esc(s.value) + '</span>' : '') +
            '<div class="text-muted">' + esc(s.note) + '</div></div>';
        });
      }
      (rec.skipped || []).forEach(s => {
        h += '<div class="small text-muted py-1">— ' + esc(s.why) + '</div>';
      });
      if (prev.payload) {
        h += '<div class="small fw-semibold mt-2">Payload SATOM would build</div>' +
          '<pre class="fw-pre small mb-1">' + esc(JSON.stringify(prev.payload, null, 2)) + '</pre>';
      }
      (prev.warnings || []).forEach(w => {
        h += '<div class="alert alert-warning py-1 px-2 small mb-1">' + esc(w) + '</div>';
      });
      (prev.errors || []).forEach(w => {
        h += '<div class="alert alert-danger py-1 px-2 small mb-1">' + esc(w) + '</div>';
      });
      if (t.explain && t.explain.summary) {
        h += '<div class="small text-muted mt-1">' + esc(t.explain.summary) + '</div>';
      }
      h += '</div>';
    });

    // --- why there is no Save button ------------------------------------
    if (d.save) {
      h += '<div class="alert alert-secondary py-2 small mb-0">' +
        '<i class="bi bi-lock me-1"></i><strong>No Save here.</strong> ' + esc(d.save.note) +
        '<div class="mt-1"><strong>' + esc(d.save.where) + '</strong></div></div>';
    }
    $('fw-fp-out').innerHTML = h;
  }

  function modalObj() {
    if (window.bootstrap && bootstrap.Modal) return bootstrap.Modal.getOrCreateInstance($(MODAL_ID));
    return null;
  }

  window.FWFpTriage = {
    open: function (prefill) {
      buildModal();
      if (prefill) { $('fw-fp-in').value = prefill; }
      const m = modalObj();
      if (m) m.show(); else $(MODAL_ID).style.display = 'block';
    }
  };

  document.addEventListener('click', function (ev) {
    const t = ev.target.closest && ev.target.closest('[data-fw-fptriage-open]');
    if (!t) return;
    ev.preventDefault();
    window.FWFpTriage.open(t.getAttribute('data-fw-fptriage-open') || '');
  });
})();

File: app/services/fp_triage.py
# app/services/fp_triage.py
"""False-positive triage from a pasted attack-log entry or raw request.

The operator's most common WAF question is not "what is this signature" — it is
*"the WAF blocked something legitimate; what exactly do I create so it stops,
and nothing else stops with it?"*. SATOM could already answer that, but ONLY
from an entry read live off a device through the attack-search panel. The entry
an operator actually has in front of them arrives in a ticket, a syslog line, a
screenshot transcription or a curl reproduction — none of which are on a device.

So this module turns pasted text into the same ``row`` dict
:mod:`app.services.attack_carveout` already consumes, and then calls that
module. There is no second recommendation engine here; a second engine is the
one that drifts.

**This path never saves, and that is a design decision, not an omission.**
``attack_carveout``'s own contract is that a carve-out is assembled from the
entry *as the device reported it*, never from values the browser sent back —
"a page that lets a client supply the evidence lets a client author the
exception". Pasted text is by definition client-supplied. So the standalone tool
explains, recommends and renders the exact payload for review, and the SAVE
action stays where the evidence is device-read: Attack Search → the entry →
Carve-out. Every response says so in :func:`triage`'s ``save`` block, because a
tool that quietly lacks a button teaches people it is broken.
"""
from __future__ import annotations

import base64
import binascii
import json
import re
from urllib.parse import unquote_plus, urlsplit

from . import attack_carveout, attack_log

#: Row keys this module will emit — taken from the attack-log column vocabulary
#: rather than typed out again, so a column added there cannot silently become
#: a field the parser drops on the floor.
ROW_KEYS = tuple(k for k, _label in attack_log.PRIMARY_FIELDS)

#: Input spellings → the canonical row key. FortiWeb's syslog, its REST payload
#: and its CSV export do not agree with each other on these names, and an
#: operator pastes whichever one their pipeline produced.
ALIASES: dict[str, str] = {
    "srcip": "src", "src_ip": "src", "source": "src", "clientip": "src",
    "dstip": "dst", "dst_ip": "dst", "destination": "dst",
    "srcport": "src_port", "dstport": "dst_port",
    "url": "http_url", "request_url": "http_url", "uri": "http_url",
    "host": "http_host", "http_hostname": "http_host",
    "method": "http_method", "http_verb": "http_method",
    "agent": "http_agent", "user_agent": "http_agent", "useragent": "http_agent",
    "referer": "http_refer", "referrer": "http_refer", "http_referer": "http_refer",
    "sigid": "signature_id", "sig_id": "signature_id",
    "signatureid": "signature_id", "signature": "signature_id",
    "subtype": "sub_type", "maintype": "main_type", "type": "main_type",
    "msgid": "msg_id", "id": "msg_id",
    "srcpolicy": "policy", "policy_name": "policy", "server_policy": "policy",
    "severity": "severity_level", "level": "severity_level",
    "threat_weight": "threat_level", "cve": "signature_cve_id",
    "owasp": "owasp_top10", "pool": "server_pool_name",
    "message": "msg", "reason": "msg",
}

#: Fields without which a specific question cannot be answered. Each entry is
#: ``(key, what it decides)`` and the missing ones are REPORTED — the operator
#: has to know which of the tool's answers rest on a field it never saw.
DECIDES: tuple[tuple[str, str], ...] = (
    ("main_type", "which WAF module produced the block, and therefore where the "
                  "exception belongs"),
    ("signature_id", "whether a per-signature exception (the narrowest fix) is "
                     "available at all"),
    ("http_url", "the path the exception is scoped to — without it most "
                 "carve-out types cannot be saved"),
    ("http_host", "which site the exception is limited to, when the profile "
                  "serves several"),
    ("policy", "which Server Policy, and therefore which Web Protection "
               "Profile, the exception has to be created on"),
)

_MAX_INPUT = 64 * 1024
_MAX_DECODE_ROUNDS = 4

_KV_RX = re.compile(r'([A-Za-z_][\w.\-]*)\s*=\s*("(?:[^"\\]|\\.)*"|\'[^\']*\'|[^\s,;]+)')
_REQLINE_RX = re.compile(
    r'^\s*(?P<m>[A-Z]{3,10})\s+(?P<u>\S+)\s+HTTP/(?P<v>[\d.]+)\s*$')


# --------------------------------------------------------------------------- #
#  Format detection + parsing                                                  #
# --------------------------------------------------------------------------- #
def _canon(key: str) -> str:
    k = str(key or "").strip().lower().replace("-", "_")
    k = ALIASES.get(k, k)
    return k


def _unquote(v: str) -> str:
    s = str(v or "").strip()
    if len(s) >= 2 and s[0] == s[-1] and s[0] in "\"'":
        s = s[1:-1]
    return s.replace('\\"', '"').replace("\\\\", "\\")


def parse_kv(text: str) -> dict:
    """FortiWeb syslog / CEF-ish ``key=value`` lines."""
    out: dict[str, str] = {}
    for m in _KV_RX.finditer(text or ""):
        out[m.group(1)] = _unquote(m.group(2))
    return out


def parse_json(text: str) -> dict:
    """A JSON object, or the first entry of the list/envelope SATOM's own attack
    log API returns. ``results``/``data`` envelopes are unwrapped because that is
    what an operator copies out of the API explorer."""
    try:
        doc = json.loads(text)
    except Exception:  # noqa: BLE001
        return {}
    for _ in range(3):
        if isinstance(doc, dict) and len(doc) and \
                any(k in doc for k in ("results", "data", "entries", "rows")):
            doc = doc.get("results") or doc.get("data") or \
                doc.get("entries") or doc.get("rows")
            continue
        break
    if isinstance(doc, list):
        doc = doc[0] if doc else {}
    if not isinstance(doc, dict):
        return {}
    return {str(k): ("" if v is None else v if isinstance(v, str) else json.dumps(v))
            for k, v in doc.items()}


def parse_http_request(text: str) -> dict:
    """A raw HTTP request — what a developer pastes from a curl reproduction.

    Deliberately produces NO ``main_type``/``signature_id``: a request carries
    what was sent, not what the appliance concluded about it. Inventing a module
    here would let the tool recommend a carve-out for a decision no device made.
    """
    lines = (text or "").replace("\r\n", "\n").split("\n")
    m = None
    start = 0
    for i, ln in enumerate(lines[:5]):
        m = _REQLINE_RX.match(ln)
        if m:
            start = i
            break
    if not m:
        return {}
    out = {"http_method": m.group("m"), "http_url": m.group("u")}
    for ln in lines[start + 1:]:
        if not ln.strip():
            break
        if ":" not in ln:
            continue
        k, v = ln.split(":", 1)
        k = k.strip().lower()
        v = v.strip()
        if k == "host":
            out["http_host"] = v
        elif k == "user-agent":
            out["http_agent"] = v
        elif k == "referer":
            out["http_refer"] = v
        elif k == "x-forwarded-for":
            out["src"] = v.split(",")[0].strip()
    return out


def parse(text: str) -> dict:
    """Turn pasted text into ``{row, fmt, mapped, unmapped, missing}``.

    ``unmapped`` is not debris to hide: a key the appliance recorded and this
    parser could not place is a field the recommendation did not see, and an
    operator staring at a narrower carve-out than they expected needs to know
    which of their evidence was not used.
    """
    raw = (text or "")[:_MAX_INPUT]
    stripped = raw.strip()
    if not stripped:
        return {"row": {}, "fmt": "", "mapped": {}, "unmapped": {},
                "missing": [d for d in DECIDES], "error": "nothing to parse"}

    fmt = ""
    flat: dict = {}
    if stripped[0] in "{[":
        flat = parse_json(stripped)
        fmt = "json" if flat else ""
    if not flat:
        req = parse_http_request(stripped)
        if req:
            flat, fmt = req, "http-request"
    if not flat:
        flat = parse_kv(stripped)
        fmt = "key=value" if flat else ""

    row: dict[str, str] = {}
    mapped: dict[str, str] = {}
    unmapped: dict[str, str] = {}
    for k, v in flat.items():
        ck = _canon(k)
        if ck in ROW_KEYS:
            if str(v).strip():
                row[ck] = str(v).strip()
                if ck != str(k).strip().lower():
                    mapped[str(k)] = ck
        else:
            unmapped[str(k)] = str(v)[:200]

    # A logged URL is often absolute. The carve-out layer wants the path and
    # already derives it, but the HOST hides in that same string and is what
    # scopes the exception to one site — so it is recovered here rather than
    # lost because the operator's pipeline folded two fields into one.
    if row.get("http_url", "").startswith(("http://", "https://")):
        parts = urlsplit(row["http_url"])
        row.setdefault("http_host", parts.netloc)

    missing = [d for d in DECIDES if not str(row.get(d[0], "")).strip()]
    return {"row": row, "fmt": fmt, "mapped": mapped, "unmapped": unmapped,
            "missing": missing,
            "error": "" if row else "no recognisable attack-log fields found"}


# --------------------------------------------------------------------------- #
#  Payload decoding                                                            #
# --------------------------------------------------------------------------- #
def decode_layers(value: str) -> list[dict]:
    """Peel URL / HTML-entity / base64 / hex encodings off a payload, in order.

    Attack-log payloads arrive encoded — usually twice — and today they are
    decoded outside SATOM, in whatever tab the operator happens to have open.
    Each round is returned separately so the operator can see WHERE the readable
    string appeared; a single fully-decoded string hides whether the appliance
    was matching on the encoded or the decoded form.
    """
    out: list[dict] = []
    cur = str(value or "")[:_MAX_INPUT]
    seen = {cur}
    for _ in range(_MAX_DECODE_ROUNDS):
        nxt, how = _decode_once(cur)
        if not nxt or nxt == cur or nxt in seen:
            break
        out.append({"how": how, "value": nxt[:4000]})
        seen.add(nxt)
        cur = nxt
    return out


def _decode_once(s: str) -> tuple[str, str]:
    if "%" in s:
        try:
            d = unquote_plus(s)
            if d != s:
                return d, "percent-decode"
        except Exception:  # noqa: BLE001
            pass
    if "&#" in s or "&lt;" in s or "&amp;" in s:
        import html
        d = html.unescape(s)
        if d != s:
            return d, "HTML entity decode"
    body = s.strip()
    if re.fullmatch(r"(?:0x)?[0-9A-Fa-f]{8,}", body) and len(body) % 2 == 0:
        try:
            d = binascii.unhexlify(body[2:] if body.lower().startswith("0x") else body)
            txt = d.decode("utf-8")
            if txt.isprintable():
                return txt, "hex decode"
        except (binascii.Error, UnicodeDecodeError, ValueError):
            pass
    if re.fullmatch(r"[A-Za-z0-9+/=_\-]{12,}", body) and len(body) % 4 in (0, 2, 3):
        try:
            d = base64.b64decode(body + "=" * (-len(body) % 4), validate=False)
            txt = d.decode("utf-8")
            # Base64 will "decode" almost anything into mojibake. Only a result
            # that is readable text is a decode; anything else is a coincidence
            # dressed as evidence.
            if txt.isprintable() and sum(c.isalnum() or c in " /<>=\"'();:,.-_" for c in txt) >= len(txt) * 0.8:
                return txt, "base64 decode"
        except (binascii.Error, UnicodeDecodeError, ValueError):
            pass
    return "", ""


# --------------------------------------------------------------------------- #
#  Triage                                                                      #
# --------------------------------------------------------------------------- #
#: Why this tool cannot save, said once so the view and the panel cannot give
#: two different accounts of the same refusal.
SAVE_NOTE = (
    "This tool explains and drafts; it does not save. A carve-out is assembled "
    "from the entry AS THE DEVICE REPORTED IT — never from values a browser "
    "sent back — so an exception authored from pasted text would let whoever "
    "supplied the text author the exception. To save this, open the same entry "
    "in Attack Search on the appliance and use the carve-out panel there: the "
    "recommendation is identical, and the evidence is device-read.")


def triage(row: dict, *, wpp: str = "") -> dict:
    """Everything SATOM can say about this entry, best carve-out first.

    Every option is proved by RUNNING the real assembly, exactly as the
    device-backed panel does — a preview produced by a second, friendlier code
    path is a preview of something that will not be what gets saved.
    """
    from . import exception_explain

    row = dict(row or {})
    labels = dict(attack_log.PRIMARY_FIELDS)
    types = attack_carveout.suggest_types(row)
    for t in types:
        t["scopers"] = [dict(s, value=str(row.get(s["row_key"]) or ""))
                        for s in attack_carveout.scopers_for(t["exc_type"])]
        subject = attack_carveout.subject_for(t["exc_type"], row)
        if subject:
            subject["label"] = labels.get(subject["row_key"], subject["row_key"])
        t["subject"] = subject
        rec = attack_carveout.recommend(row, t["exc_type"])
        rec["preview"] = attack_carveout.build(row, t["exc_type"], rec["picked"])
        t["recommended"] = rec
        t["explain"] = exception_explain.explain(
            t["exc_type"], rec["preview"].get("payload") or {},
            wpp=wpp, policy=row.get("policy", ""))
    return {
        "types": types,
        "policy": row.get("policy", ""),
        "wpp": wpp,
        "save": {"can_save": False, "note": SAVE_NOTE,
                 "where": "Attack Search → the entry → Carve-out"},
    }


__all__ = ["ROW_KEYS", "ALIASES", "DECIDES", "SAVE_NOTE",
           "parse", "parse_kv", "parse_json", "parse_http_request",
           "decode_layers", "triage"]
